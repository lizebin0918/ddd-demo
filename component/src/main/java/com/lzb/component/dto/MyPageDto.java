package com.lzb.component.dto;

import java.util.List;
import java.util.function.Function;

import lombok.AllArgsConstructor;
import lombok.Getter;
import one.util.streamex.StreamEx;

/**
 * <br/>
 * Created on : 2023-09-22 23:07
 * @author lizebin
 */
@Getter
@AllArgsConstructor
public class MyPageDto<T> {

    private final long pages;
    private final long pageSize;
    private final long total;
    private final List<T> rows;

    private <E> MyPageDto(long pages, long pageSize, long total, List<E> rows, Function<E, T> convertor) {
        this.pages = pages;
        this.pageSize = pageSize;
        this.total = total;
        this.rows = StreamEx.of(rows).map(convertor).toList();
    }

    public static <E, T> MyPageDto<T> of(long pages,
            long pageSize,
            long total,
            List<E> rows,
            Function<E, T> convertor) {
        return new MyPageDto<>(pages, pageSize, total, rows, convertor);
    }
}
