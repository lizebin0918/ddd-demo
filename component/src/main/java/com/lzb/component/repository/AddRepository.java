package com.lzb.component.repository;


import com.lzb.component.aggregate.BaseEntity;
import jakarta.validation.constraints.NotNull;

import org.springframework.validation.annotation.Validated;

/**
 * 新增聚合<br/>
 * Created on : 2022-03-10 09:59
 *
 * @author lizebin
 */
@Validated
public interface AddRepository<R extends BaseEntity<I>, I> {

    /**
     * 新增聚合根
     * @param aggregate
     */
    I add(@NotNull(message = "聚合根不能为空") R aggregate);

}
