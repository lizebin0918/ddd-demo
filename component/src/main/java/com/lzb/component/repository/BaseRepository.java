package com.lzb.component.repository;

import java.util.Objects;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

import com.lzb.component.aggregate.BaseAggregation;
import com.lzb.component.event.DomainEventSupport;
import com.lzb.component.helper.TransactionHelper;
import jakarta.annotation.Resource;
import lombok.NonNull;
import lombok.Setter;

import org.springframework.aop.framework.AopContext;
import org.springframework.context.ApplicationContext;
import org.springframework.context.ApplicationContextAware;

@Setter
public abstract class BaseRepository<R extends BaseAggregation<R, I>, I> implements CommonRepository<R, I>, ApplicationContextAware {

    @Resource
    protected TransactionHelper transactionHelper;

    @Resource
    protected DomainEventSupport domainEventSupport;

    private ApplicationContext applicationContext;

    @Override
    public void setApplicationContext(ApplicationContext applicationContext) {
        this.applicationContext = applicationContext;
    }

    /**
     * 添加聚合根
     *
     * @param aggregate
     * @return 返回主键
     */
    public abstract Supplier<I> doAdd(R aggregate);

    /**
     * 更新聚合根
     *
     * @param aggregate
     * @return
     */
    public abstract Runnable doUpdate(R aggregate);

    protected BaseRepository<R, I> getCurrentProxy() {
        if (Objects.isNull(applicationContext)) {
            return this;
        }
        return (BaseRepository<R, I>) AopContext.currentProxy();
    }

    @Override
    public I add(@NonNull R aggregate) {
        AtomicReference<I> id = new AtomicReference<>();
        Supplier<I> idSupplier = getCurrentProxy().doAdd(aggregate);
        transactionHelper.runWithRequired(() -> {
            id.set(idSupplier.get());
            domainEventSupport.sendEventAfterPersist(aggregate.getEvents());
        });
        return id.get();
    }


    @Override
    public void update(@NonNull R aggregate) {
        Runnable doUpdate = getCurrentProxy().doUpdate(aggregate);
        transactionHelper.runWithRequired(() -> {
            doUpdate.run();
            domainEventSupport.sendEventAfterPersist(aggregate.getEvents());
        });
    }
}