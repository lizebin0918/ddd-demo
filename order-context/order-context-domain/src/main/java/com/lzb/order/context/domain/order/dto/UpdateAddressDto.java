package com.lzb.order.context.domain.order.dto;

import jakarta.validation.constraints.NotBlank;

/**
 * 更新订单地址<br/>
 * Created on : 2023-09-16 10:29
 * @author lizebin
 */
public record UpdateAddressDto(long orderId,
                               @NotBlank
                                    String email,
                               @NotBlank
                                    String phoneNumber, String firstName,
                               String lastName, String addressLine1, String addressLine2, String country) {

}
