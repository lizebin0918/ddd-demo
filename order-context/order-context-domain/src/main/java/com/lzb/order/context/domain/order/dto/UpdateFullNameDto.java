package com.lzb.order.context.domain.order.dto;

import jakarta.validation.constraints.NotBlank;

/**
 * 前端只会传姓名，只更新姓名即可<br/>
 * Created on : 2023-09-17 21:34
 * @author lizebin
 */
public record UpdateFullNameDto(
        long orderId,
        @NotBlank
        String firstName,
        String lastName) {

}
