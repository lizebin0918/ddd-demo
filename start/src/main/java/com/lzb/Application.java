package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.autoconfigure.data.redis.RedisRepositoriesAutoConfiguration;
import org.springframework.cache.annotation.EnableCaching;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.scheduling.annotation.EnableAsync;

// 启动报错：Multiple Spring Data modules found, entering strict repository configuration
@SpringBootApplication(exclude = { RedisRepositoriesAutoConfiguration.class})
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
