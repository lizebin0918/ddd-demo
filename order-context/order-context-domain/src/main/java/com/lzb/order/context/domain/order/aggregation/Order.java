package com.lzb.order.context.domain.order.aggregation;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.function.IntUnaryOperator;
import java.util.stream.Collectors;

import cn.hutool.core.lang.Assert;
import com.lzb.component.aggregate.BaseAggregation;
import com.lzb.component.exception.BizException;
import com.lzb.order.context.domain.core.valobj.SkuCount;
import com.lzb.order.context.domain.inventory.entity.SkuStockLock;
import com.lzb.order.context.domain.order.aggregation.valobj.FullAddressLine;
import com.lzb.order.context.domain.order.aggregation.valobj.FullName;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderAddress;
import com.lzb.order.context.domain.order.aggregation.valobj.OrderStatus;
import com.lzb.order.context.domain.order.dto.UpdateAddressDto;
import com.lzb.order.context.domain.order.event.OrderCanceledEvent;
import com.lzb.order.context.domain.order.event.OrderPlacedEvent;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
@Jacksonized
@SuperBuilder(toBuilder = true)
@Setter(AccessLevel.PACKAGE)
public class Order extends BaseAggregation<Order, Long> {

    private Long id;

    private OrderStatus orderStatus;

    private String currency;

    private BigDecimal exchangeRate;

    private BigDecimal totalShouldPay;

    private BigDecimal totalActualPay;

    private OrderAddress orderAddress;

    private OrderDetails orderDetails;

    /**
     * 扩展字段
     */
    private Extension extension;

    @Builder.Default
    private Long userId = 1L;

    @NotNull
    public Set<Integer> getSkuIds() {
        return skuCounts().stream().map(SkuCount::skuId).collect(Collectors.toSet());
    }

    public void updateAddress(UpdateAddressDto updateOrderAddress) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        this.orderAddress = OrderAddress.builder()
            .fullName(FullName.of(updateOrderAddress.firstName(), updateOrderAddress.lastName()))
            .fullAddressLine(FullAddressLine.of(updateOrderAddress.addressLine1(), updateOrderAddress.addressLine2()))
            .country(updateOrderAddress.country())
            .email(updateOrderAddress.email())
            .phoneNumber(updateOrderAddress.phoneNumber())
            .build();
    }

    public void updateSkuLockStock(List<SkuStockLock> skuStockLocks) {
        IntUnaryOperator skuId2LockNum = lockedNum(skuStockLocks);
        orderDetails.forEach(orderDetail -> {
            orderDetail.updateLocked(orderDetail.getCount() <= skuId2LockNum.applyAsInt(orderDetail.getSkuId()));
        });
    }

    private IntUnaryOperator lockedNum(List<SkuStockLock> skuStockLocks) {
        Map<Integer, SkuStockLock> skuId2Locks = skuStockLocks.stream().collect(Collectors.toMap(SkuStockLock::skuId, Function.identity()));
        return skuId -> Optional.ofNullable(skuId2Locks.get(skuId)).map(SkuStockLock::lockedNum).orElse(0);
    }

    public void pay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = orderStatus.toCancel();
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return orderStatus.isCancel();
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        Assert.notBlank(fullName.getFirstName(), "firstName is null");
        this.orderAddress = orderAddress.toBuilder().fullName(fullName).build();
    }

    public boolean canCancel() {
        return false;
    }

    public void addOrderDetail(@NonNull OrderDetail orderDetail) {
        if (Objects.isNull(orderDetails)) {
            this.orderDetails = new OrderDetails(new ArrayList<>());
        }
        Set<Integer> inSkuIds = orderedSkuIds();
        if (inSkuIds.contains(orderDetail.getSkuId())) {
            throw new BizException("skuId重复");
        }
        orderDetails.add(orderDetail);
    }

    /**
     * 下单的skuIds
     * @return
     */
    private @NotNull Set<Integer> orderedSkuIds() {
        return orderDetails.toStream()
                .map(OrderDetail::getSkuId)
                .collect(Collectors.toSet());
    }

    public List<SkuCount> skuCounts() {
        return orderDetails.toStream().map(item -> new SkuCount(item.getSkuId(), item.getCount())).toList();
    }

    @Override
    public Long id() {
        return id;
    }
}
