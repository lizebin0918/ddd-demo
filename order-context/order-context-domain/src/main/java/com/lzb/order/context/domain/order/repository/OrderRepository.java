package com.lzb.order.context.domain.order.repository;

import java.util.List;
import java.util.Optional;

import com.lzb.component.repository.CacheRepository;
import com.lzb.component.repository.CommonRepository;
import com.lzb.order.context.domain.order.aggregation.Order;
import com.lzb.order.context.domain.order.aggregation.OrderReadonly;

/**
 * <br/>
 * Created on : 2023-08-30 23:07
 * @author lizebin
 */
public interface OrderRepository extends CommonRepository<Order, Long>, CacheRepository<Order, Long> {

    List<Order> list(long... orderIds);

    default Optional<OrderReadonly> getReadonly(long orderId) {
        return get(orderId).map(OrderReadonly::of);
    }

    default Optional<OrderReadonly> getReadonlyInCache(long orderId) {
        return getInCache(orderId).map(OrderReadonly::of);
    }
}
