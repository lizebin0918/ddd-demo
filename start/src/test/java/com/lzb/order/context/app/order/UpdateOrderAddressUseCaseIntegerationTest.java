package com.lzb.order.context.app.order;

import com.lzb.BaseIntegrationTest;
import com.lzb.order.context.domain.order.dto.UpdateAddressDto;
import com.lzb.order.context.domain.order.dto.UpdateFullNameDto;
import com.lzb.order.context.domain.order.aggregation.Order;
import com.lzb.order.context.domain.order.repository.OrderRepository;
import jakarta.annotation.Resource;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.test.context.jdbc.Sql;

class UpdateOrderAddressUseCaseIntegerationTest extends BaseIntegrationTest {

    @Resource
    private OrderAppService orderAppService;

    @Resource
    private OrderRepository orderRepository;

    @Test
    @DisplayName("测试订单地址修改")
    @Sql("/sql/UpdateOrderAddressUseCaseIntegerationTest/should_update_order_address.sql")
    void should_update_order_address() {
        long orderId = 1L;
        UpdateAddressDto updateAddressDto = new UpdateAddressDto(orderId, "email1", "phoneNumber1", "firstName1",
                "lastName1", "addressLine11", "addressLine21", "country1");
        orderAppService.updateAddress(updateAddressDto);

        Order order = orderRepository.getOrThrow(orderId);
        assertJSON(order.getOrderAddress());
    }

    @Test
    @DisplayName("测试姓名修改")
    @Sql("/sql/UpdateOrderAddressUseCaseIntegerationTest/should_update_full_name.sql")
    void should_update_full_name() {
        long orderId = 1L;

        UpdateFullNameDto cmd = new UpdateFullNameDto(orderId, "li", "");
        orderAppService.updateFullName(cmd);

        Order order = orderRepository.getOrThrow(orderId);
        assertJSON(order.getOrderAddress());
    }

}