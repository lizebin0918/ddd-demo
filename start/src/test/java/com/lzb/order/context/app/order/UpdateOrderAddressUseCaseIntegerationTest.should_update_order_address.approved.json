{
  "country": "country1",
  "email": "email1",
  "fullAddressLine": {
    "addressLine1": "addressLine11",
    "addressLine2": "addressLine21"
  },
  "fullName": {
    "firstName": "firstName1",
    "lastName": "lastName1"
  },
  "phoneNumber": "phoneNumber1"
}